File: java/array_search/BitonicArraySearch.java
public class BitonicArraySearch {

    private static int findPeak(int[] arr) {
        int left = 0;
        int right = arr.length - 1;

        while (left < right) {
            int mid = left + (right - left) / 2;

            if (arr[mid] < arr[mid + 1]) left = mid + 1;
            else right = mid;
        }

        return left;
    }

    private static int binarySearchIncreasing(int[] arr, int target, int left, int right) {
        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (arr[mid] == target) return mid;
            if (arr[mid] < target) left = mid + 1;
            else right = mid - 1;
        }

        return -1;
    }

    private static int binarySearchDecreasing(int[] arr, int target, int left, int right) {
        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (arr[mid] == target) return mid;
            if (arr[mid] > target) left = mid + 1;
            else right = mid - 1;
        }

        return -1;
    }

    public static int bitonicSearch(int[] arr, int target) {
        if (arr.length == 0) return -1;

        int peak = findPeak(arr);
        int index = binarySearchIncreasing(arr, target, 0, peak);

        if (index != -1) return index;

        return binarySearchDecreasing(arr, target, peak + 1, arr.length - 1);
    }
}

File: python/array_search/bitonic_array_search.py
def _find_peak(arr):
    left = 0
    right = len(arr) - 1

    while left < right:
        mid = (left + right) // 2
        if arr[mid] < arr[mid + 1]:
            left = mid + 1
        else:
            right = mid

    return left


def _binary_search_increasing(arr, target, left, right):
    while left <= right:
        mid = (left + right) // 2

        if arr[mid] == target:
            return mid
        if arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1

    return -1


def _binary_search_decreasing(arr, target, left, right):
    while left <= right:
        mid = (left + right) // 2

        if arr[mid] == target:
            return mid
        if arr[mid] > target:
            left = mid + 1
        else:
            right = mid - 1

    return -1


def bitonic_search(arr, target):
    if not arr:
        return -1

    peak = _find_peak(arr)
    index = _binary_search_increasing(arr, target, 0, peak)

    if index != -1:
        return index

    return _binary_search_decreasing(arr, target, peak + 1, len(arr) - 1)
